### DISCUSSION TITLE
vweb: implement a default graceful shutdown

### DISCUSSION THREAD

**Discussion by smartiniOnGitHub** [vlang/v#11765] [Ideas]
Hi all,
to successfully close a vweb application, it would be great to have a default method called when the application get a SIGTERM signal  (maybe even SIGINT); note that could be useful to call that handler even by application code (when applications expose a termination endpoint on a protected route or port, or other).
Default behavior should be to stop incoming traffic and stop the server (using a timeout, by default x msec); this is useful for example to let existing elaborations to complete, and flush DB connections (if any).
And if an application need a custom logic, write a custom close handler.

Note that an internal flag to manage incoming traffic could be useful even when webapps are running for example in Kubernetes (k8s), where it's possible to specify even a "readiness" endpoint that when true k8s will route incoming traffic to the webapp; so for example it could be automatically set to true at the end of webapp init (the "init_once" method) and and the beginning of the close handler.
All just to give more context ... that in some cases the ability to stop (maybe temporarily) application to accept/process incoming requests could be good to have; and in those cases maybe by default reply with an HTTP error 503, as seen [here](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/503).

What do you think ?

Thanks a lot

**Comment by smartiniOnGitHub** [comment]
I just  saw an interesting approach to graceful shutdown in a Go framework, using a channel of os signals, maybe could be implemented in a similar way even here, see [here](https://medium.com/swlh/single-binary-web-apps-in-go-and-vue-part-1-ea7d4100eab7). Bye

**Comment by smartiniOnGitHub** [comment]
Hi all, I've been able to do some tests on this topic; as you can see in the following example (for a simple cli application), inspired by 'vwatch.v' source:
```
module main

import os
import time

// get input (interactive) from console, 
// until the process receives a termination signal

// defines an execution context for the current app
struct Context {
mut:
	pid        int  // the pid of the current process; useful while debugging
	is_exiting bool // set by SIGINT/Ctrl-C
	lines      []string
}

// instantiation of execution context, as a constant
const ccontext = Context{}

// register_signal_handler register an handler function for most common closing signals
// note that signals SIGKILL and SIGSTOP cannot be caught, blocked, or ignored
fn register_signal_handler(ctx Context) {
    // println(@FN + ' Got ${ctx.lines.len} lines') // expect 0 here
    assert ctx.lines.len == 0

    /*
    // later check how to handle more signals in the same statement ...
	// os.signal(C.SIGINT | C.SIGTERM, fn () { // with or it doesn't work
	os.signal(C.SIGINT, fn () {
		mut ctx := &ccontext
		ctx.is_exiting = true
        println(@FN + ' Got ${ctx.lines.len} lines') // expect some lines here (after some user input)
        exit(1)
	})
     */
    // TODO: comment previous block (working), then call dedicated function/method ... ok
	os.signal(C.SIGINT,  close_handler)
	os.signal(C.SIGTERM, close_handler)
}

// close_handler manage the closing of the application, via a catchable signal
fn close_handler() {
    println(@FN + ' Got closing signal ...')

    mut ctx := &ccontext
    ctx.is_exiting = true
    println(@FN + ' Got ${ctx.lines.len} lines') // expect some lines here (after some user input)
	time.sleep(1 * time.second)  // add a delay before exiting, as a sample

    // close the current process, otherwise common terminating signals will be ignored
    exit(ctx.lines.len) // close the current process, as a sample with a different exit code based on input
    // TODO: check if there is a better/cleaner way to close the current application ... wip
}

fn main() {
    println('---- Start to get input (interactive) from console,  until the process receives a termination signal ---')

    // get a mutable reference to execution context
	mut context := &ccontext
	context.pid = os.getpid()
    println('Current process ID (PID): ${context.pid}')

    register_signal_handler(context) // pass context as argument to the function
    // later check if implement register_signal_handler as a method of context ...

    // loop forever
    for {
        context.lines << os.input('Write something and press <ENTER>: ')
    }

    println('---- End ---') // never executed
}
```
I think a similar approach could be used even with vweb, pass application instance to the closing handler to be able to close open resources etc ... do you think it could be a good way (so I can start to draft a PR) ?
Thanks for the feedback and the help. Bye

**Comment by smartiniOnGitHub** [comment]
@mcastorina what do you think ?

**Comment by mcastorina** [comment]
Hm.. I see no reason why that couldn't technically work, however I think it is better design to keep signal handler logic short.

I would prefer spawning the server in a thread, then notifying the thread to gracefully shutdown, but that's my opinion.

Alternative idea: change `run` and `run_app` to return a `Result` type and handle shutdown operations in the `or {}` block

```v
vweb.run<App>(port) or {
    println('shutting down..')
}
```

**Comment by smartiniOnGitHub** [comment]
ok, thanks for the feedback; so I think that for this feature it's better to wait the rework of vweb in multi-thread.
My only concern here is that we should manage resources that are not directly tied to user sessions (for example a DB connection pool), so not sure App is the right place ... let's look at this later, thanks again. Bye
